p2p/src/run_p2p.rs

```diff
@@ -0,0 +1,385 @@
+    }
+
+    async fn get_connected_peers(&self) -> crate::Result<Vec<ConnectedPeer>> {
+        let (tx, rx) = oneshot_nofail::channel();
+        self.peer_manager_mbox_tx
+            .send(PeerManagerEvent::GetConnectedPeers(tx))
+            .map_err(P2pError::from)?;
+        rx.await.map_err(P2pError::from)
+    }
+
+    async fn add_reserved_node(&mut self, addr: String) -> crate::Result<()> {
+        let addr = addr
+            .parse::<T::Address>()
+            .map_err(|_| P2pError::ConversionError(ConversionError::InvalidAddress(addr)))?;
+        self.peer_manager_mbox_tx
+            .send(PeerManagerEvent::AddReserved(addr))
+            .map_err(|_| P2pError::ChannelClosed)?;
+        Ok(())
+    }
+
+    async fn remove_reserved_node(&mut self, addr: String) -> crate::Result<()> {
+        let addr = addr
+            .parse::<T::Address>()
+            .map_err(|_| P2pError::ConversionError(ConversionError::InvalidAddress(addr)))?;
+        self.peer_manager_mbox_tx
+            .send(PeerManagerEvent::RemoveReserved(addr))
+            .map_err(|_| P2pError::ChannelClosed)?;
+        Ok(())
+    }
+
+    async fn submit_transaction(&mut self, tx: SignedTransaction) -> crate::Result<()> {
+        let id = tx.transaction().get_id();
+        self.mempool_handle.call_mut(|m| m.add_transaction(tx)).await??;
+        self.messaging_handle.broadcast_message(SyncMessage::NewTransaction(id))
+    }
+
+    fn subscribe_to_events(
+        &mut self,
+        handler: Arc<dyn Fn(P2pEvent) + Send + Sync>,
+    ) -> crate::Result<()> {
+        self.p2p_event_handlers_tx.send(handler).map_err(Into::into)
+    }
+}
+
+mod utils {
+
+    use std::{future::Future, time::Duration};
+
+    use futures::{future::BoxFuture, never::Never, FutureExt};
+    use logging::log;
+    use tokio::sync::oneshot;
+
+    use crate::Result;
+
+    type Killswitch = oneshot::Sender<Never>;
+
+    pub fn spawn_with_killswitch<F>(f: F) -> (BoxFuture<'static, Result<()>>, Killswitch)
+    where
+        F: Future<Output = Result<()>> + Send + 'static,
+    {
+        let (ks_tx, ks_rx) = killswitch_chan();
+        let jh = tokio::spawn(async move {
+            tokio::select! {
+                _ = ks_rx => Err(crate::P2pError::Cancelled),
+                outcome = f => outcome
+            }
+        });
+        (jh.map(|r| r.expect("join failure")).fuse().boxed(), ks_tx)
+    }
+
+    pub async fn run_multiple<SS, KS, F>(
+        mut services: Vec<F>,
+        mut shutdown_switches: Vec<Option<(SS, KS, Duration)>>,
+    ) -> Result<()>
+    where
+        F: Future<Output = Result<()>> + Unpin,
+    {
+        assert_eq!(services.len(), shutdown_switches.len());
+
+        log::trace!("run-multiple [len: {}]", services.len());
+
+        let (outcome, idx, _rest) = futures::future::select_all(services.iter_mut()).await;
+
+        if let Err(reason) = &outcome {
+            log::warn!("Service[{}] shut down with error: {}", idx, reason);
+        }
+
```

Comment:
The idea is that the tasks with higher indices in the tasks-vec depend on those that have the lower indices.
Therefore if a task terminates, everything "to the right" of it has little chances to work correctly, so it is shut down harshly.

The tasks "to the left" are asked to shut down nicely sequentially.